src/nbd.c

```diff
@@ -0,0 +1,1286 @@
+
+	g_message("configuring nbd device");
+
+	nl = netlink_connect(&driver_id);
+	if (!nl) {
+		res = FALSE;
+		g_set_error(error, G_IO_ERROR, G_IO_ERROR_FAILED, "failed to connect to netlink");
+		goto out;
+	}
+
+	msg = nlmsg_alloc();
+	if (!msg)
+		g_error("failed to allocate netlink message");
+
+	if (!genlmsg_put(msg, NL_AUTO_PORT, NL_AUTO_SEQ, driver_id, 0, 0, NBD_CMD_CONNECT, 0))
+		g_error("failed to add generic netlink headers to message");
+
+	NLA_PUT_U64(msg, NBD_ATTR_SIZE_BYTES, nbd_dev->data_size);
+	NLA_PUT_U64(msg, NBD_ATTR_BLOCK_SIZE_BYTES, 4096);
+	NLA_PUT_U64(msg, NBD_ATTR_SERVER_FLAGS, 0);
+	NLA_PUT_U64(msg, NBD_ATTR_CLIENT_FLAGS,
+			NBD_CFLAG_DESTROY_ON_DISCONNECT |
+			NBD_CFLAG_DISCONNECT_ON_CLOSE
+			);
+	NLA_PUT_U64(msg, NBD_ATTR_TIMEOUT, 300);
+
+	attr_sockets = nla_nest_start(msg, NBD_ATTR_SOCKETS);
+	if (!attr_sockets)
+		g_error("failed to allocate nested NBD_ATTR_SOCKETS netlink message");
+	attr_item = nla_nest_start(msg, NBD_SOCK_ITEM);
+	if (!attr_item)
+		g_error("failed to allocate nested NBD_SOCK_ITEM netlink message");
+	NLA_PUT_U32(msg, NBD_SOCK_FD, nbd_dev->sock);
+	nla_nest_end(msg, attr_item);
+	nla_nest_end(msg, attr_sockets);
+
+	nl_socket_modify_cb(nl, NL_CB_VALID, NL_CB_CUSTOM, netlink_connect_cb, nbd_dev);
+	if (nl_send_sync(nl, msg) < 0) {
+		res = FALSE;
+		g_set_error(error, G_IO_ERROR, G_IO_ERROR_FAILED, "netlink send_sync failed");
+		goto out;
+	}
+	if (!nbd_dev->index_valid)
+		g_error("failed to create nbd device");
+
+	nbd_dev->dev = g_strdup_printf("/dev/nbd%"G_GUINT32_FORMAT, nbd_dev->index);
+
+	g_message("setup done for %s", nbd_dev->dev);
+
+	res = TRUE;
+	goto out;
+
+nla_put_failure:
+	res = FALSE;
+	g_set_error(error, G_IO_ERROR, G_IO_ERROR_FAILED, "netlink NLA_PUT failed");
+
+out:
+	if (nl) {
+		nl_close(nl);
+		nl_socket_free(nl);
+	}
+	return res;
+}
+
+gboolean remove_nbd_device(RaucNBDDevice *nbd_dev, GError **error)
+{
+	gboolean res = FALSE;
+	int driver_id;
+	struct nl_sock *nl = NULL;
+	struct nl_msg *msg = NULL;
+
+	g_return_val_if_fail(nbd_dev != NULL, FALSE);
+
+	if (!nbd_dev->index_valid)
+		return TRUE;
+
+	g_message("remove_nbd_device");
+
+	nl = netlink_connect(&driver_id);
+	if (!nl) {
+		res = FALSE;
+		g_set_error(error, G_IO_ERROR, G_IO_ERROR_FAILED, "failed to connect to netlink");
+		goto out;
+	}
+
+	msg = nlmsg_alloc();
+	if (!msg)
+		g_error("failed to allocate netlink message");
+
+	if (!genlmsg_put(msg, NL_AUTO_PORT, NL_AUTO_SEQ, driver_id, 0, 0, NBD_CMD_DISCONNECT, 0))
+		g_error("failed to add generic netlink headers to message");
+
+	NLA_PUT_U32(msg, NBD_ATTR_INDEX, nbd_dev->index);
+
+	if (nl_send_sync(nl, msg) < 0) {
+		g_message("nbd device is already removed");
+	}
+
+	nbd_dev->index_valid = FALSE;
+	nbd_dev->index = 0;
+
+	/* maybe reuse the socket to get final statistics/error message? */
+	g_close(nbd_dev->sock, NULL);
+	nbd_dev->sock = -1;
+	g_clear_pointer(&nbd_dev->dev, g_free);
+
+	res = TRUE;
+	goto out;
+
+nla_put_failure:
```

Comment:
Nitpick, I would add a comment that this is used by the `NLA_PUT_*` family of macros.

Reply:
Good idea.